ipaserver/install/ipa_idrange_fix.py

```diff
@@ -0,0 +1,1073 @@
+            metavar=200000,
+            help="Threshold for a gap between out-of-range IDs to be \
+considered a different range. Has to be > 0",
+        )
+        parser.add_option(
+            "--minrange",
+            dest="minrange",
+            type=int,
+            default=10,
+            metavar=10,
+            help="Minimal considered range size for out-of-range IDs.\
+All ranges with amount of IDs lower than this number will be discarded and \
+IDs will be listed to be moved manually. Has to be > 1",
+        )
+        parser.add_option(
+            "--allowunder1000",
+            dest="allowunder1000",
+            action="store_true",
+            default=False,
+            help="Allow idranges to start below 1000. Be careful to not \
+overlap IPA users/groups with existing system-local ones!",
+        )
+        parser.add_option(
+            "--norounding",
+            dest="norounding",
+            action="store_true",
+            default=False,
+            help="Disable IDrange rounding attempt in order to get ranges \
+exactly covering just IDs provided",
+        )
+        parser.add_option(
+            "--unattended",
+            dest="unattended",
+            action="store_true",
+            default=False,
+            help="Automatically fix all range issues found without asking \
+for confirmation",
+        )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.realm: str = None
+        self.suffix: DN = None
+        self.propositions_rid: list[IDRange] = []
+        self.propositions_new: list[IDRange] = []
+        self.outliers: list[IDentity] = []
+        self.under1000: list[IDentity] = []
+        self.id_ranges: list[IDRange] = []
+
+    def validate_options(self, needs_root=True):
+        # super(IPAIDRangeFix, self).validate_options(needs_root)
+        super().validate_options(needs_root)
+
+    def run(self):
+        api.bootstrap(in_server=True)
+        api.finalize()
+
+        self.realm = api.env.realm
+        self.suffix = ipautil.realm_to_suffix(self.realm)
+        try:
+            api.Backend.ldap2.connect()
+
+            # Reading range data
+            self.id_ranges = read_ranges(self.suffix)
+
+            # Evaluating existing ranges, if something is off, exit
+            if self.evaluate_ranges() != 0:
+                return 1
+
+            # reading out of range IDs
+            ids_out_of_range = read_outofrange_identities(
+                self.suffix, self.id_ranges
+            )
+
+            # Evaluating out of range IDs
+            self.evaluate_identities(ids_out_of_range)
+
+            # Print the propositions
+            self.print_intentions()
+
+            # If there are no propositions, we have nothing to do, exiting
+            if (len(self.propositions_rid) == 0
+                    and len(self.propositions_new) == 0):
+                logger.info("\nNo changes proposed, nothing to do.")
```

Comment:
Is this none proposed or none required? Should it be "no changes needed" or something like that?

Reply:
That's a good question. Changes the tool can propose are either adding RID bases or creating new ranges. If we have some users that didn't fit the criteria for new ranges creation, changes are needed, albeit manual ones, but the tool cannot help with those. In this part of code, we don't know if there are any, though, hence this ambiguous wording.